extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/WorkerImpl.java

```diff
@@ -394,6 +433,17 @@ private void handleNewWorkOrder(
     kernelHolder.partitionBoundariesFuture = runWorkOrder.getStagePartitionBoundariesFuture();
   }
 
+  @SuppressWarnings("unchecked")
+  private void updateMetricDimensions(QueryDefinition queryDef)
+  {
+    Set<String> datasourceDim = (Set<String>) queryMetricDimensions.computeIfAbsent(DruidMetrics.DATASOURCE, s -> new HashSet<>());
+    Set<Interval> intervalDim = (Set<Interval>) queryMetricDimensions.computeIfAbsent(DruidMetrics.INTERVAL, s -> new HashSet<>());
+    for (StageDefinition stageDef : queryDef.getStageDefinitions()) {
+      datasourceDim.addAll(MSQMetricUtils.getDatasources(stageDef));
```

Comment:
nit: is this a one time need - or maybe there should be `queryDef.getDatasources`  ?

Reply:
Currently, I believe this is a one time need. Now that I have changed getDatasources to a member of stageDefinition, maybe this isn't required.